File: depth-first-search/diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int diameterOfBinaryTree(TreeNode root) {
        //because arrays are call by reference, u can use this instead of a global variable
        int max[] = new int[1];
        maxDepth(root, max);
        return max[0];
    }
    private int maxDepth(TreeNode root, int[] max){
        if(root == null)
            return 0;
        int left = maxDepth(root.left, max);
        int right = maxDepth(root.right, max);

        max[0] = Math.max(left+right, max[0]);

        return 1 + Math.max(left,right);
    }
}

File: binary-tree/diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    int diameter = 0;
public:
    int dfs(TreeNode* root){
        if(root == NULL) return 0;

        int leftHeight = 0, rightHeight = 0, currDia = 0;
        leftHeight = dfs(root->left);
        rightHeight = dfs(root->right);
        currDia = leftHeight + rightHeight;

        diameter = max(diameter, currDia);

        return 1 + max(leftHeight, rightHeight);
    }
    int diameterOfBinaryTree(TreeNode* root) {
        dfs(root);
        return diameter;
    }
};